### PR TITLE
pt: explicitly set device

### PR DESCRIPTION
**Pull request by njzjz**
Forcely requires setting the device to `env.DEVICE` or `cpu` explicitly for functions like `torch.tensor`, `torch.zeros`, `torch.ones`, `torch.rand`, `torch.eye`, `LayerNorm`, and data loader.
This ensures that no OP runs on the wrong device. The trick here is `torch.set_default_device("cuda:9999999")` in the tests, so errors will be thrown if the default device is used.

Tips:
(1) Avoid `torch.zeros(...).to(device=...)`. This first initializes memory on CPUs and copies it to GPUs.
(2) Use `with torch.device(...)` for a module that cannot set a device (i.e., the data loader).

### CODE REVIEW

**Review by wanghan-iapcm — APPROVED** [review]

File: deepmd/pt/train/training.py [hunk 1]
```diff
@@ -519,7 +521,8 @@ def step(_step_id, task_key="Default"):
                     if not torch.isfinite(grad_norm).all():
                         # check local gradnorm single GPU case, trigger NanDetector
                         raise FloatingPointError("gradients are Nan/Inf")
-                self.optimizer.step()
+                with torch.device("cpu"):
+                    self.optimizer.step()
```
Comment (iProzd) [inline deepmd/pt/train/training.py:525]: why use device settings here?

File: deepmd/pt/utils/dataloader.py [hunk 2]
```diff
@@ -139,8 +139,9 @@ def construct_dataset(system):
             self.total_batch += len(system_dataloader)
         # Initialize iterator instances for DataLoader
         self.iters = []
-        for item in self.dataloaders:
-            self.iters.append(iter(item))
+        with torch.device("cpu"):
+            for item in self.dataloaders:
+                self.iters.append(iter(item))
```
Comment (iProzd) [inline deepmd/pt/utils/dataloader.py:144]: Does this device setting matter here?

**Review by anyangml — APPROVED** [review]

File: deepmd/pt/train/training.py (hunk 1 above)
Comment (njzjz) [reply deepmd/pt/train/training.py:525]: https://github.com/pytorch/pytorch/blob/c28a43988e7f3c630e5c50717c146c2cd936eabf/torch/optim/adam.py#L104-L111

File: deepmd/pt/utils/dataloader.py (hunk 2 above)
Comment (njzjz) [reply deepmd/pt/utils/dataloader.py:144]: https://github.com/pytorch/pytorch/blob/c28a43988e7f3c630e5c50717c146c2cd936eabf/torch/utils/data/dataloader.py#L602

Comment (iProzd) [reply deepmd/pt/utils/dataloader.py:144]: got it

**Review by iProzd — APPROVED** [review]

**Pull request merged**